### PR TITLE
Add CLI for x/data

### LINKED ISSUE #131: [x/data] add CLI commands
Create cli commands for data module functionality (both `tx` and `query` functionality, mirroring MsgServer and QueryServers)

### LINKED ISSUE #152: x/data CLI query command: Use `by-cid` by default
**Is your feature request related to a problem? Please describe.**

Right now, to query for a cid in x/data, we need to do:
```bash
regen q data by-cid <cid>
```

**Describe the solution you'd like**

```bash
regen q data <cid>
```

If <cid> is a subcommand, run that subcommand. If not, run `QueryByCidCmd`.

**Describe alternatives you've considered**

**Additional context**

ref: https://github.com/regen-network/regen-ledger/pull/147#issuecomment-722416463

### PR DESCRIPTION
**Pull request by amaurymartiny**
Fixes #131
fixes #152 

## Proposed Changes
  - Add CLI commands for Msg/{Sign,Anchor,Store}
  - Rename the `Data` query to `ByCid` to be more explicit

TODO:
- [x] Run manually the commands.

Feel free to submit a Draft Pull Request as soon as you start working on something. Before you mark your PR as Ready For Review, make sure you've checked off the following boxes or indicated N/A (not applicable):
- [x] linked to a relevant issue or have written a clear description of the issue in the PR
- [ ] wrote or updated tests
- [ ] wrote or updated documentation
- [ ] added an appropriate entry in `CHANGELOG.md` following https://keepachangelog.com

Thanks!

### CODE REVIEW

File: x/data/client/tx.go [hunk 1]
```diff
@@ -25,22 +31,119 @@ func TxCmd() *cobra.Command {
...
+				return err
+			}
+
+			cid, err := gocid.Decode(args[0])
+			if err != nil {
+				return err
+			}
+
+			msg := data.MsgSignDataRequest{
+				Signers: []string{clientCtx.GetFromAddress().String()},
```
Comment (amaurymartiny) [inline x/data/client/tx.go:92]: I'm still not sure what's the best UX for CLI users. Say 3 signers want to sign the same CID:
- either they run this CLI command each individually
- or this CLI command just generates a tx file, and people pass around that tx file and each signs the file with their CLI (via x/auth's Sign command)

I prefer the first UX, so put it here for now.

File: x/data/client/query.go [hunk 2]
```diff
@@ -25,6 +25,7 @@ func QueryCmd() *cobra.Command {
 	return cmd
 }
 
+// QueryDataCmd created a CLI command for Query/DATA.
```
Comment (amaurymartiny) [inline x/data/client/query.go:28]:
```suggestion
// QueryDataCmd creates a CLI command for Query/DATA.
```

File: x/data/client/tx.go [hunk 3]
```diff
@@ -25,22 +31,119 @@ func TxCmd() *cobra.Command {
...
+
+	flags.AddTxFlagsToCmd(cmd)
+
+	return cmd
+}
+
+// MsgStoreDataCmd creates a CLI command for Msg/StoreData.
+func MsgStoreDataCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "store [cid] [content-as-base64]",
```
[the remark this answers is not in the document]
Comment (anilCSE) [reply x/data/client/tx.go]: JSON string would be handy, else I would prefer reading from a file

**Comment by amaurymartiny** [comment]
Running the CLI commands manually:
- `regen tx data anchor <key_name> <cid>`: works
- `regen q data data <cid>`: works
- `regen tx data sign <key_name> <cid>`: Got error `"wrong number of signers; expected [ DEB891DC7A0D39F4BBE2F59807EDA87199D096C4], got 1: unauthorized"`

I'm investigating this error right now.

**Comment by aaronc** [comment]
This is a bit awkward: `regen q data data <cid>`. Can we just do `regen q data ...`?

**Comment by amaurymartiny** [comment]
> This is a bit awkward: regen q data data <cid>. Can we just do regen q data ...?

What if we want to add more queries to x/data in the future?

We could rename the 2nd `data` to e.g. `get-cid`

**Comment by aaronc** [comment]
Yeah something like that or `by-cid`. Can a root cmd so something and still have sub commands though? Just `query data` seems convenient.

**Comment by amaurymartiny** [comment]
> Can a root cmd so something and still have sub commands though?

I took 5min to look at Cobra, didn't find anything. Maybe someone with more experience with cobra has an idea? In any case, I created https://github.com/regen-network/regen-ledger/issues/152

**Comment by aaronc** [comment]
Pretty sure it will work. See how gov does it with submit proposal: x/gov/client/cli/tx.go

File: proto/regen/data/v1alpha1/query.proto [hunk 4]
```diff
@@ -9,20 +9,20 @@ option go_package = "github.com/regen-network/regen-ledger/x/data";
 // Query is the regen.data.v1alpha1 Query service
 service Query {
 
-    // Data queries data based on its CID.
-    rpc Data(QueryDataRequest) returns (QueryDataResponse);
+    // ByCid queries data based on its CID.
+    rpc ByCid(QueryByCidRequest) returns (QueryByCidResponse);
```
Comment (amaurymartiny) [inline proto/regen/data/v1alpha1/query.proto:13]: This change is not necessarily needed, but I think it's clearer, esp given that the module name is already data

**Comment by amaurymartiny** [comment]
R4R.

File: util/msg_service.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
+package util
```
Comment (robert-zaremba) [inline util/msg_service.go:1]: Let's rename the package and directory to `/client`.

File: x/data/requests.go [hunk 6]
```diff
@@ -26,12 +26,12 @@ func (m *MsgSignDataRequest) ValidateBasic() error {
...
 	addrs := make([]sdk.AccAddress, len(m.Signers))
 
-	for _, signer := range m.Signers {
+	for i, signer := range m.Signers {
 		addr, err := sdk.AccAddressFromBech32(signer)
 		if err != nil {
 			panic(err)
 		}
-		addrs = append(addrs, addr)
+		addrs[i] = addr
```
Comment (robert-zaremba) [inline x/data/requests.go:34]: :+1:

**Review by robert-zaremba — DISMISSED** [review]
let's rename the `util` package

File: x/data/client/tx.go [hunk 7]
```diff
@@ -20,27 +26,141 @@ func TxCmd() *cobra.Command {
...
+				return err
+			}
+
+			msg := data.MsgSignDataRequest{
+				Signers: []string{clientCtx.GetFromAddress().String()},
+				Cid:     cid.Bytes(),
+			}
+			svcMsgClientConn := &util.ServiceMsgClientConn{}
+			msgClient := data.NewMsgClient(svcMsgClientConn)
+			_, err = msgClient.SignData(context.Background(), &msg)
```
Comment (robert-zaremba) [inline x/data/client/tx.go:109]: We shouldn't use `context.Background`. Let's use `cmd.Context()`

File: x/data/client/tx.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (clevinson) [reply x/data/client/tx.go]: For now i think base64 is fine. We can also provide some examples in the tutorial for encoding json or files like...
```
cat ecological_report.pdf | base64 | regen store [sender] [cid]
```

Another thought, maybe in the future we don't actually need cid provided with the `store` command? We should be able determine the cid internally in the function ya? Is there any added value in requiring the end-user to calculate the cid on their own beforehand?

File: util/msg_service.go (hunk 5 above)
Comment (clevinson) [reply util/msg_service.go]: ack

File: x/data/client/tx.go (hunk 7 above)
Comment (clevinson) [reply x/data/client/tx.go]: ack

**Comment by clevinson** [comment]
Addressed @robert-zaremba's comments. Can you mark as approved so we can merge?

**Review by clevinson — APPROVED** [review]

**Pull request merged**